ENH: add anatomical label finding from atlas to dipole fit tutorial

mdovgialo opened the pull request:
I needed to find dipole labels from anatomical atlas, couldn't find any info how to do it until I stumbled on MNE discord and larsoner  helped me.

We both agree this should be in tutorial, or maybe there should be even a function/helper to find anatomical labels easily.

It might be worth looking at https://github.com/miykael/atlasreader/tree/master/atlasreader integration.

drammock commented on tutorials/inverse/20_dipole_fit.py:
```diff
@@ -57,8 +59,29 @@
+best_dip_id = dip.gof.argmax()
+best_dip_mri_pos = mri_pos[best_dip_id]
+
+# Load parcelization:
```
```suggestion
# Load parcellation:
```

drammock commented on tutorials/inverse/20_dipole_fit.py:
```diff
@@ -57,8 +59,29 @@
+
+# Load parcelization:
+atlas_fname = op.join(subjects_dir, subject, 'mri', 'aparc.a2009s+aseg.mgz')
+parcelation_img = nibabel.load(atlas_fname)
```
```suggestion
parcellation_img = nibabel.load(atlas_fname)
```

drammock commented on tutorials/inverse/20_dipole_fit.py:
```diff
@@ -57,8 +59,29 @@
+label_lut = {v: k for k, v in lut_inv_dict.items()}
+
+# Find voxel for dipole position
+mri_vox_t = np.linalg.inv(parcelation_img.header.get_vox2ras_tkr())
```
```suggestion
mri_vox_t = np.linalg.inv(parcellation_img.header.get_vox2ras_tkr())
```

drammock commented on tutorials/inverse/20_dipole_fit.py:
```diff
@@ -57,8 +59,29 @@
+vox_dip_pos = mne.transforms.apply_trans(mri_vox_t, best_dip_mri_pos).astype(int)
+
+# Get voxel value and label from LUT
+vol_values = parcelation_img.get_fdata()[tuple(vox_dip_pos.T)]
```
```suggestion
vol_values = parcellation_img.get_fdata()[tuple(vox_dip_pos.T)]
```

drammock commented on tutorials/inverse/20_dipole_fit.py:
```diff
@@ -57,8 +59,29 @@
+vol_values = parcelation_img.get_fdata()[tuple(vox_dip_pos.T)]
+label = label_lut.get(vol_values, 'Unknown')
+
+# Draw the dipole position on the MRI scan and write the anatomical label from parcelization
```
```suggestion
# Draw dipole position on MRI scan and add anatomical label from parcellation
```
(each line must be <80 characters)

drammock commented on tutorials/inverse/20_dipole_fit.py:
```diff
@@ -57,8 +59,29 @@
+# Draw the dipole position on the MRI scan and write the anatomical label from parcelization
 t1_fname = op.join(subjects_dir, subject, 'mri', 'T1.mgz')
-fig_T1 = plot_anat(t1_fname, cut_coords=mri_pos[0], title='Dipole loc.')
+fig_T1 = plot_anat(t1_fname, cut_coords=mri_pos[0], title='Dipole loc. anatomical label: {}'.format(label))
```
```suggestion
fig_T1 = plot_anat(t1_fname, cut_coords=mri_pos[0],
                   title=f'Dipole location: {label}')
```

agramfort commented on tutorials/inverse/20_dipole_fit.py:
```diff
@@ -57,8 +59,31 @@
+
+# Get voxel value and label from LUT
+vol_values = parcellation_img.get_fdata()[tuple(vox_dip_pos.T)]
+label = label_lut.get(vol_values, 'Unknown')
```
I find this code a bit scary for average users and this usecase seems quite common.

how about we add a

`find_label_in_annot(pos, subject='sample', annot="a2009s+aseg")`

function?

wdyt?

mdovgialo replied:
Yes it is a good idea, I'm just a little busy with my projects and wanted to start with a small contribution which would be helpful.
More like `find_label_in_annot(pos, subject='sample', annot="a2009s+aseg", subject_dir=dir)`

But where to put such function? In what package? What test case to add and where? Just to test a few known positions?

agramfort commented:
>
> More like find_label_in_annot(pos, subject='sample', annot="a2009s+aseg",
> subject_dir=dir)
>

yes !


> But where to put such function? In what package? What test case to add and
> where? Just to test a few known positions?
>

in mne/label.py and the test in mne/tests/test_label.py

no rush. thanks !

mdovgialo commented:
@agramfort @drammock 
Something like this maybe?

drammock commented on mne/label.py:
```diff
@@ -2666,3 +2669,43 @@ def select_sources(subject, label, location='center', extent=0.,
     return new_label
+
+
+def find_label_in_annot(pos, subject='fsaverage', annot="aparc.a2005s+aseg",
```
```suggestion
def find_label_in_annot(pos, subject='fsaverage', annot='aparc.a2005s+aseg',
```

drammock commented on mne/label.py:
```diff
@@ -2666,3 +2669,43 @@ def select_sources(subject, label, location='center', extent=0.,
+
+    Parameters
+    ----------
+    pos : vector of x,y,z coordinates in MRI space
+    subject : MRI subject name
+    annot : parcellation file name
+    subjects_dir : path to MRI subjects directory
```
see how we do parameters in our other public functions. First line of each entry is `param name : allowed type | other allowed type`. Second and subsequent lines are indented with description formatted like a proper sentence/paragraph.

drammock commented on mne/label.py:
```diff
@@ -2666,3 +2669,43 @@ def select_sources(subject, label, location='center', extent=0.,
+    # Find voxel for dipole position
+    mri_vox_t = np.linalg.inv(parcellation_img.header.get_vox2ras_tkr())
+    vox_dip_pos_f = mne.transforms.apply_trans(mri_vox_t, pos)
+    vox_dip_pos = vox_dip_pos_f.astype(int)
```
might want `np.rint()` here?  `.astype(int)` will turn a coordinate of `5.99` into `5`

drammock commented on mne/tests/test_label.py:
```diff
@@ -13,15 +13,17 @@
                            assert_equal)
 import pytest
 
+import mne
```
we don't `import mne` in the test files; just import the specific functions you need

drammock commented on mne/tests/test_label.py:
```diff
@@ -13,15 +13,17 @@
-                       select_sources)
-from mne.utils import requires_sklearn, get_subjects_dir, check_version
+                       select_sources, find_label_in_annot)
+from mne.utils import requires_sklearn, get_subjects_dir, check_version, \
+    requires_nibabel
```
```suggestion
from mne.utils import (requires_sklearn, get_subjects_dir, check_version,
                       requires_nibabel)
```

drammock commented on mne/tests/test_label.py:
```diff
@@ -1015,3 +1017,20 @@ def test_select_sources():
+@testing.requires_testing_data
+@requires_nibabel()
+def test_find_label_in_annot():
+    """Test searching for atlass name for given MRI position."""
```
```suggestion
    """Test searching for atlas name for given MRI position."""
```

drammock reviewed (changes requested):
this is looking pretty good. Note that I didn't have time to play around with it / fully test it out, just read the code.

drammock commented on mne/label.py:
```diff
@@ -9,16 +9,19 @@
                               _center_of_mass, extract_label_time_course,
                               spatial_src_adjacency)
-from .source_space import add_source_space_distances, SourceSpaces
+from .source_space import add_source_space_distances, SourceSpaces, \
+    read_freesurfer_lut
```
```suggestion
from .source_space import (add_source_space_distances, SourceSpaces,
                           read_freesurfer_lut)
```

drammock commented on mne/label.py:
```diff
@@ -2666,3 +2669,48 @@ def select_sources(subject, label, location='center', extent=0.,
+
+    Parameters
+    ----------
+    pos : instance of np.array
```
```suggestion
    pos : ndarray, shape (3,)
```

drammock commented on mne/label.py:
```diff
@@ -2666,3 +2669,48 @@ def select_sources(subject, label, location='center', extent=0.,
+        MRI subject name.
+    annot : str
+        MRI parcellation file name.
+    subjects_dir : str
```
```suggestion
    subjects_dir : path-like
```

drammock commented on mne/label.py:
```diff
@@ -2666,3 +2669,48 @@ def select_sources(subject, label, location='center', extent=0.,
+    subject : str
+        MRI subject name.
+    annot : str
+        MRI parcellation file name.
```
```suggestion
        MRI parcellation file name. Do not include the ``.mgz`` suffix.
```

drammock reviewed (approved):
awesome. just a couple tiny things.

mdovgialo commented:
Ready to merge, I guess?

drammock commented:
I'm satisfied, I'll let @agramfort or @larsoner merge if they're happy.

Greendayle commented:
Ooops, seems like I managed to overlook some style issue, I'm falling asleep, I'll fix it when I wake up

mdovgialo commented:
Fixed the small style issue

drammock commented:
Just realized this probably should get an entry in the enhancements section of our changelog (`doc/changes/latest.inc` and corresponding entry in `names.inc`). Sorry I didn't think of that when reviewing... I'm always forgetting to do it for my own changes too.

mdovgialo commented:
> 
> 
> Just realized this probably should get an entry in the enhancements section of our changelog (`doc/changes/latest.inc` and corresponding entry in `names.inc`). Sorry I didn't think of that when reviewing... I'm always forgetting to do it for my own changes too.

Like this?

drammock commented:
> > Just realized this probably should get an entry in the enhancements section of our changelog (`doc/changes/latest.inc` and corresponding entry in `names.inc`). Sorry I didn't think of that when reviewing... I'm always forgetting to do it for my own changes too.
> 
> Like this?

Yep, looks good.

mdovgialo commented:
@drammock 
Docs failed to build for some reason and I have no idea why:

```
embedding documentation hyperlinks for auto_examples... [ 97%] ssp_projs_sensitivity_map.html
embedding documentation hyperlinks for auto_examples... [ 98%] topo_compare_conditions.html
embedding documentation hyperlinks for auto_examples... [ 99%] topo_customized.html
embedding documentation hyperlinks for auto_examples... [100%] xhemi.html

Added  32 HTML plot_* redirects for auto_tutorials
Added 109 HTML plot_* redirects for auto_examples
Added  43 HTML custom redirects
ERROR:mprof:Program resulted with a non-zero exit code: 1
make: *** [Makefile:65: html_dev-pattern-memory] Error 1
make: Leaving directory '/home/circleci/project/doc'
```

Please help

drammock commented:
line 1815 of the docs build log: `changes/latest.inc:38: WARNING: py:func reference target not found: mne.label.find_label_in_annot`

You'll need to add the new function to one of the API pages for cross-references to it to work.  (`doc/source_space.rst` is probably the right place to put it)

drammock commented:
```
WARNING: [numpydoc] Validation warnings while processing docstring for 'mne.label.find_label_in_annot':
  GL03: Double line break found; please use only one blank line to separate sections or paragraphs, and do not leave blank lines at the end of docstrings
  RT03: Return value has no description
```

mdovgialo commented:
>  RT03: Return value has no description

I'm confused by this one. But maybe this will fix it.

larsoner commented on mne/label.py:
```diff
@@ -2666,3 +2669,49 @@ def select_sources(subject, label, location='center', extent=0.,
     return new_label
+
+
+def find_label_in_annot(pos, subject='fsaverage', annot='aparc.a2005s+aseg',
```
The default for `read_labels_from_annot` is `aparc` so this should be `aparc`. Also, nothing in this actually has to do with `mne.Label` objects, so let's rename
```suggestion
def find_pos_in_annot(pos, subject='fsaverage', annot='aparc+aseg',
```

larsoner replied:
... I just realized what I think you meant by the original name, which is "Find the label in annot corresponding to a given pos". But I read it as "Find where a given mne.Label is in annot". Overall I think "find_pos_in_annot" being short for "Find where pos is in annot" makes the most sense here, especially since what you get back might be the name of a mne.Label (not a label object itself) or it might be the name of a region that you would use with a volumetric atlas.

larsoner commented on mne/label.py:
```diff
@@ -2666,3 +2669,49 @@ def select_sources(subject, label, location='center', extent=0.,
+    # Find voxel for dipole position
+    mri_vox_t = np.linalg.inv(parcellation_img.header.get_vox2ras_tkr())
+    vox_dip_pos_f = apply_trans(mri_vox_t, pos)
+    vox_dip_pos = np.rint(vox_dip_pos_f).astype(int)
```
```suggestion
    vox_dip_pos = np.rint(vox_dip_pos_f, dtype=int)
```

larsoner commented on mne/label.py:
```diff
@@ -2666,3 +2669,49 @@ def select_sources(subject, label, location='center', extent=0.,
+    -----
+    .. versionadded:: 0.24
+    """
+    import nibabel
```
Give our standard error message:
```suggestion
    nibabel = _import_nibabel('read MRI parcellations')
```
and above import `_import_nibabel` from `.utils`

larsoner commented on mne/label.py:
```diff
@@ -2666,3 +2669,49 @@ def select_sources(subject, label, location='center', extent=0.,
+        subjects_dir = get_subjects_dir(None)
+    atlas_fname = os.path.join(subjects_dir, subject, 'mri', annot + '.mgz')
+    parcellation_img = nibabel.load(atlas_fname)
+
```
Need some checking here:
```
    pos = np.asarray(pos, float)
    if pos.shape != (3,):
        raise ValueError('pos must be an array of shape (3,), '
                         f'got {pos.shape}')
```

larsoner commented on mne/label.py:
```diff
@@ -2666,3 +2669,49 @@ def select_sources(subject, label, location='center', extent=0.,
+    subject : str
+        MRI subject name.
+    annot : str
+        MRI parcellation file name. Do not include the ``.mgz`` suffix.
```
Eventually we probably want to add support for nearest-surface-label functionality using the `lh.aparc.annot` (and similar files) and the cortical ribbon, so for now let's make it clear that it's going to use a volumetric atlas only at this point:
```suggestion
        MRI volumetric atlas file name.
        Do not include the ``.mgz`` suffix.
```

larsoner commented on doc/changes/latest.inc:
```diff
@@ -35,6 +35,7 @@ Enhancements
 
 - Add support for NIRSport devices to `mne.io.read_raw_nirx` (:gh:`9348` **by new contributor** |David Julien|_, **new contributor** |Romain Derollepot|_, `Robert Luke`_, and `Eric Larson`_)
 
+- New function :func:`mne.label.find_label_in_annot` to get atlas label for MRI coordinates.  (:gh:`9376` by `Marian Dovgialo`_)
```
```suggestion
- New function :func:`mne.label.find_label_in_annot` to get atlas label for MRI coordinates.  (:gh:`9376` **by new contributor** |Marian Dovgialo|_)
```
And above you'll need to add a bold-name-substitution like we have for David Julien et al.

larsoner commented on doc/source_space.rst:
```diff
@@ -42,3 +42,4 @@ Source Space Data
    write_labels_to_annot
    write_label
    source_space.compute_distance_to_sensors
+   label.find_label_in_annot
```
```suggestion
   label.find_pos_in_annot
```

larsoner reviewed (commented):
Looks great, just a few minor comments!

larsoner commented on mne/label.py:
```diff
@@ -9,16 +9,19 @@
 
 import numpy as np
 
+from mne.transforms import apply_trans
```
all module imports should be nested (and preferably alphabetical), will push a commit to fix

larsoner reviewed (approved):
Otherwise LGTM!

larsoner commented:
Thanks @mdovgialo !
